File: python/lca.py
max_depth = 1
tree = dict()
depth = []
parent = []

def init(cur, p):
    depth[cur] = depth[p] + 1
    parent[cur][0] = p

    for i in range(1, max_depth + 1):
        parent[cur][i] = parent[parent[cur][i - 1]][i - 1]

    for nxt in tree[cur]:
        if nxt == p:
            continue
        init(nxt, cur)

def query(a, b):
    if a == 1 or b == 1:
        return 1
    
    if depth[a] > depth[b]:
        a, b = b, a

    if depth[a] != depth[b]:
        for i in range(max_depth, -1, -1):
            if depth[parent[b][i]] >= depth[a]:
                b = parent[b][i]
    
    lca = a
    if a != b:
        for i in range(max_depth, -1, -1):
            if parent[a][i] != parent[b][i]:
                a = parent[a][i]
                b = parent[b][i]
            lca = parent[a][i]

    return lca

File: rust/src/lca.rs
use std::collections::VecDeque;
use std::mem::swap;

struct Tree {
    tree: Vec<Vec<usize>>,
    parent: Vec<Vec<usize>>,
    depth: Vec<usize>,
}

impl Tree {
    fn new(size: usize, tree: Vec<Vec<usize>>) -> Self {
        Self { 
            tree, 
            parent: vec![vec![0; 17]; size + 1], 
            depth: vec![0; size + 1],
        }
    }

    fn init_parent(&mut self, root: usize) {
        let mut q = VecDeque::new();
        self.depth[root] = 1;
        self.parent[root][0] = 0;
        q.push_back(root);

        while let Some(node) = q.pop_front() {
            for &child in self.tree[node].iter() {
                if child != self.parent[node][0] {
                    self.depth[child] = self.depth[node] + 1;
                    self.parent[child][0] = node;
                    q.push_back(child);
                }
            }
        }
    }

    fn fill_parent(&mut self) {
        for j in 1..17 {
            for i in 1..self.tree.len() {
                self.parent[i][j] = self.parent[self.parent[i][j - 1]][j - 1];
            }
        }
    }

    fn query(&self, mut a: usize, mut b: usize) -> usize {
        if self.depth[a] < self.depth[b] {
            swap(&mut a, &mut b);
        }

        let diff = self.depth[a] - self.depth[b];

        for i in 0..17 {
            if diff & (1 << i) != 0 {
                a = self.parent[a][i];
            }
        }

        if a != b {
            for i in (0..17).rev() {
                if self.parent[a][i] != self.parent[b][i] {
                    a = self.parent[a][i];
                    b = self.parent[b][i];
                }
            }
            a = self.parent[a][0];
        }

        a
    }
}